File: dotnet/src/MyDotey.SCF/Filter/DefaultValueFilter.cs
using System;

namespace MyDotey.SCF.Filter
{
    /**
     * @author koqizhao
     *
     * May 21, 2018
     * 
     * apply each filter to the value,
     * if null returned by a filter, stop filtering and return null,
     * otherwise, apply next filter to the value
     */
    public class DefaultValueFilter<V> : AbstractValueFilter<V>
    {
        private Func<V, V> _filter;

        public DefaultValueFilter(Func<V, V> filter)
        {
            if (filter == null)
                throw new ArgumentNullException("filter is null");
            _filter = filter;
        }

        public override V Filter(V t)
        {
            return _filter(t);
        }

        public override int GetHashCode()
        {
            int prime = 31;
            int result = 1;
            result = prime * result + ((_filter == null) ? 0 : _filter.GetHashCode());
            return result;
        }

        public override bool Equals(Object obj)
        {
            if (!base.Equals(obj))
                return false;

            DefaultValueFilter<V> other = (DefaultValueFilter<V>)obj;
            return object.Equals(_filter, other._filter);
        }

        public override string ToString()
        {
            return string.Format("{0} {{ filter: {1} }}", GetType().Name, _filter);
        }
    }
}

File: java/src/main/java/org/mydotey/scf/filter/DefaultValueFilter.java
package org.mydotey.scf.filter;

import org.mydotey.java.ObjectExtension;

public class DefaultValueFilter<T> extends AbstractValueFilter<T> {

    private T _defaultValue;

    public DefaultValueFilter(T defaultValue) {
        ObjectExtension.requireNonNullOrEmpty(defaultValue, "defaultValue");
        _defaultValue = defaultValue;
    }

    @Override
    public T apply(T t) {
        return ObjectExtension.isNullOrEmpty(t) ? _defaultValue : t;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((_defaultValue == null) ? 0 : _defaultValue.hashCode());
        return result;
    }

    @SuppressWarnings("rawtypes")
    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        DefaultValueFilter other = (DefaultValueFilter) obj;
        if (_defaultValue == null) {
            if (other._defaultValue != null)
                return false;
        } else if (!_defaultValue.equals(other._defaultValue))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "DefaultValueFilter [defaultValue=" + _defaultValue + "]";
    }

}
